consensus/polybft/consensus_runtime.go

```diff
@@ -0,0 +1,851 @@
+func (c *consensusRuntime) getStateSyncEventsForBundle(from, bundleSize uint64) ([]*StateSyncEvent, error) {
+	until := bundleSize + from
+
+	return c.state.getStateSyncEventsForCommitment(from, until)
+}
+
+// startEventTracker starts the event tracker that listens to state sync events
+func (c *consensusRuntime) startEventTracker() error {
+	if c.eventTracker != nil {
+		return nil
+	}
+
+	c.eventTracker = &eventTracker{
+		config:     c.config.PolyBFTConfig,
+		subscriber: c,
+		dataDir:    c.config.DataDir,
+		logger:     c.logger.Named("event_tracker"),
+	}
+
+	if err := c.eventTracker.start(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// deliverMessage receives the message vote from transport and inserts it in state db for given epoch.
+// It returns indicator whether message is processed successfully and error object if any.
+func (c *consensusRuntime) deliverMessage(msg *TransportMessage) (bool, error) {
+	epoch := c.getEpoch()
+	if epoch == nil || msg.EpochNumber < epoch.Number {
+		// Epoch metadata is undefined
+		// or received message for some of the older epochs.
+		return false, nil
+	}
+
+	if !c.isActiveValidator() {
+		return false, fmt.Errorf("validator is not among the active validator set")
+	}
+
+	// check just in case
+	if epoch.Validators == nil {
+		return false, fmt.Errorf("validators are not set for the current epoch")
+	}
+
+	msgVote := &MessageSignature{
+		From:      msg.NodeID,
+		Signature: msg.Signature,
+	}
+
+	if err := validateVote(msgVote, epoch); err != nil {
+		return false, err
+	}
+
+	numSignatures, err := c.state.insertMessageVote(msg.EpochNumber, msg.Hash, msgVote)
+	if err != nil {
+		return false, fmt.Errorf("error inserting message vote: %w", err)
+	}
+
+	c.logger.Info(
+		"deliver message",
+		"hash", hex.EncodeToString(msg.Hash),
+		"sender", msg.NodeID,
+		"signatures", numSignatures,
+		"quorum", getQuorumSize(len(epoch.Validators)),
+	)
+
+	return true, nil
+}
+
+func (c *consensusRuntime) runCheckpoint(epoch *epochMetadata) error {
+	// TODO: Implement checkpoint
+	return nil
+}
+
+// getLatestSprintBlockNumber returns latest sprint block number
+func (c *consensusRuntime) getLatestSprintBlockNumber() uint64 {
+	lastBuiltBlockNumber := c.lastBuiltBlock.Number
+
+	sprintSizeMod := lastBuiltBlockNumber % c.config.PolyBFTConfig.SprintSize
+	if sprintSizeMod == 0 {
+		return lastBuiltBlockNumber
+	}
+
+	sprintBlockNumber := lastBuiltBlockNumber - sprintSizeMod
+
+	return sprintBlockNumber
+}
+
+// calculateUptime calculates uptime for blocks starting from the last built block in current epoch,
+// and ending at the last block of previous epoch
+func (c *consensusRuntime) calculateUptime(currentBlock *types.Header) (*UptimeCounter, error) {
+	epoch := c.getEpoch()
+	uptimeCounter := &UptimeCounter{validatorIndices: make(map[ethgo.Address]int)}
+
+	if c.config.PolyBFTConfig.EpochSize < (uptimeLookbackSize + 1) {
+		// this means that epoch size must at least be 3 blocks,
```

Comment:
Is this a hard limit, having the epoch be of size 3 (minimum)?

cc @Kourin1996

Reply:
Yep, it is a limitation due to the following two things:
1. we are checking parent signatures at the current block
2. last block of the epoch includes update of validator set, so when calculating uptime we need to skip the last block, as it already contains validators for the next epoch.

Anyhow this is going to be changed in the future once we have slashing implemented (which would require dynamic epoch sizes).

cc @goran-ethernal @ferranbt